src/crimson/os/seastore/extent_placement_manager.h

```diff
@@ -0,0 +1,193 @@
+
+namespace extent_placement_manager {
+  struct segment_header_t {
+    std::string magic;
+    DENC(segment_header_t, v, p) {
+      DENC_START(1, 1, p);
+      denc(v.magic, p);
+      DENC_FINISH(p);
+    }
+  };
+}
+
+class ExtentRewriter {
+public:
+  using write_iertr = trans_iertr<crimson::errorator<
+    crimson::ct_error::input_output_error, // media error or corruption
+    crimson::ct_error::invarg,             // if offset is < write pointer or misaligned
+    crimson::ct_error::ebadf,              // segment closed
+    crimson::ct_error::enospc              // write exceeds segment size
+    >>;
+
+  virtual write_iertr::future<> write(std::list<LogicalCachedExtentRef>& extent) = 0;
+  virtual ~ExtentRewriter() {}
+};
+
+class SegmentedRewriter : public ExtentRewriter,
+                          public boost::intrusive_ref_counter<
+  SegmentedRewriter, boost::thread_unsafe_counter>{
+public:
+  using roll_segment_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  using alloc_extent_ertr = roll_segment_ertr;
+  using init_segment_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+
+  SegmentedRewriter(SegmentProvider& sp, SegmentManager& sm)
+    : segment_provider(sp), segment_manager(sm) {}
+  write_iertr::future<> write(std::list<LogicalCachedExtentRef>& extent) final;
+
+private:
+  bool _needs_roll(segment_off_t length) const;
+  roll_segment_ertr::future<> roll_segment();
+  init_segment_ertr::future<> init_segment(Segment& segment);
+
+  using extents_to_write_t = std::vector<LogicalCachedExtentRef>;
+  void add_extent_to_write(bufferlist& bl,
+                           extents_to_write_t& extents_to_write,
+                           LogicalCachedExtentRef& extent);
+
+  SegmentProvider& segment_provider;
+  SegmentManager& segment_manager;
+  SegmentRef current_segment;
+  std::vector<SegmentRef> open_segments;
+  segment_off_t allocated_to = 0;
+};
+
+using SegmentedRewriterRef = std::unique_ptr<SegmentedRewriter>;
+
+class ExtentAllocator {
+public:
+  using scan_device_ertr = crimson::errorator<
+    crimson::ct_error::input_output_error>;
+  virtual CachedExtentRef alloc_ool_extent(
+    extent_types_t type,
+    const CachedExtentRef&,
+    segment_off_t length) = 0;
+
+  virtual scan_device_ertr::future<> scan_device() = 0;
+  virtual ~ExtentAllocator() {};
+};
+
+using ExtentAllocatorRef = std::unique_ptr<ExtentAllocator>;
+
+class SegmentedAllocator : public ExtentAllocator {
+public:
+  using alloc_extent_ertr = SegmentedRewriter::alloc_extent_ertr;
+  SegmentedAllocator(
+    SegmentProvider& sp,
+    SegmentManager& sm,
+    Cache& cache)
+    : segment_provider(sp),
+      segment_manager(sm),
+      cache(cache)
+  {}
+
+  CachedExtentRef alloc_ool_extent(
+    extent_types_t type,
+    const CachedExtentRef& old_extent,
+    segment_off_t length) final;
+
+
+  template <typename T>
+  TCachedExtentRef<T> alloc_ool_extent(
+    const CachedExtentRef& old_extent,
```

Comment:
This is unused except for calc_target_merit which also doesn't use it.  I don't think this interface should be restricted to rewriting an extent, remove this param.